### PR TITLE
query-frontend: Added support of auto_discovery for memcached

### LINKED ISSUE #6979: query-frontend: does not support auto_discovery with memcached
**Thanos, Prometheus and Golang version used**:

```
$ thanos --version
thanos, version 0.33.0-dev (branch: main, revision: 2ec9b0caef685246908c592526fb024ea9369b0e)
  build date:       20231214-14:31:39
  go version:       go1.21.4
  platform:         linux/amd64
  tags:             netgo
```

**Object Storage Provider**:
doesn't matter

**What happened**:
According to documetation https://thanos.io/tip/components/query-frontend.md/#memcached
thanos query uses the same memcached config as thanos store. So it should support `auto_discovery` option.
Instead query-fronted tries to do get/put operations at autodiscovery address and doesn't even try to discover nodes.

**What you expected to happen**:
I expected that config with `auto_discovery: true` would work.

**How to reproduce it (as minimally and precisely as possible)**:
run thanos query-frontend with memcached with `auto_discovery: true` and address pointing to a setup supporting auto_discovery.

Look at the code: https://github.com/thanos-io/thanos/blob/v0.32.5/pkg/queryfrontend/config.go#L134-L151

It unmarshals config properly, but then it composes a new config and discards any info about auto_discovery

**Full logs to relevant components**:

I have local auto_discovery service at `localhost:11211`.
it works:
```
$ telnet localhost 11211
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
test
CONFIG cluster 0 68
1
10.235.84.55|10.235.84.55|11211 10.234.203.59|10.234.203.59|11211

Connection closed by foreign host.
```
thanos query-frontend tries to put into autodiscovery address instead of discovered nodes
```
ts=2023-12-14T14:46:08.11474845Z caller=memcached.go:250 level=error msg="failed to put to memcached" name= err="write tcp 127.0.0.1:53946->127.0.0.1:11211: write: connection reset by peer"
```

It works OK with thanos store.

**Anything else we need to know**:

**My questions and suggestions**:

Is it intentional?

Should I dig deeper and find a way to make it work with auto_discovery?

### PR DESCRIPTION
**Pull request by xBazilio**
* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

Fixes #6979

## Verification

Tested with local auto discovery server. Ran `thanos query-frontend` with memcached config setting `"addresses":[localhost:11211], "auto_discovery": true`

### CODE REVIEW

File: internal/cortex/chunk/cache/memcached_client.go [hunk 1]
```diff
@@ -68,6 +71,7 @@ type MemcachedClientConfig struct {
 	Host           string        `yaml:"host"`
 	Service        string        `yaml:"service"`
 	Addresses      string        `yaml:"addresses"` // EXPERIMENTAL.
+	AutoDicovery   bool          `yaml:"auto_dicovery"`
```
Comment (TaylorMutch) [inline internal/cortex/chunk/cache/memcached_client.go:74]: Should this be AutoDiscovery?

Comment (xBazilio) [reply internal/cortex/chunk/cache/memcached_client.go]: thanx, fixed.

File: CHANGELOG.md [hunk 2]
```diff
@@ -31,6 +31,7 @@ We use *breaking :warning:* to mark changes that are not backward compatible (re
 
 ### Fixed
 
+- [#7004](https://github.com/thanos-io/thanos/pull/7004) Query Frontend: support documented auto discovery for memcached
```
Comment (yeya24) [inline CHANGELOG.md:34]: Can we add this to `Added` section?

File: internal/cortex/chunk/cache/memcached_client.go [hunk 3]
```diff
@@ -107,9 +111,19 @@ func NewMemcachedClient(cfg MemcachedClientConfig, name string, r prometheus.Reg
...
 	client.MaxIdleConns = cfg.MaxIdleConns
 
-	dnsProviderRegisterer := prometheus.WrapRegistererWithPrefix("cortex_", prometheus.WrapRegistererWith(prometheus.Labels{
-		"name": name,
-	}, r))
+	var addressProvider clientconfig.AddressProvider
+	if cfg.AutoDiscovery {
+		addressProvider = memcacheDiscovery.NewProvider(
+			logger,
+			extprom.WrapRegistererWithPrefix("thanos_memcached_", r),
```
Comment (yeya24) [inline internal/cortex/chunk/cache/memcached_client.go:118]: Is it better to use the same metric prefix here? `cortex_`

Comment (xBazilio) [reply internal/cortex/chunk/cache/memcached_client.go]: It was copied from thanos store code. I did as you suggested.

File: CHANGELOG.md (hunk 2 above)
Comment (xBazilio) [reply CHANGELOG.md]: done

**Review by yeya24 — APPROVED** [review]

**Pull request merged**